File: wk4/java/shiritori.java
// problem link: https://open.kattis.com/problems/shiritori

import java.io.*;
import java.math.*;
import java.util.*;

public class shiritori {
    public static void main(String[] args) throws IOException {
        BufferedReader r = new BufferedReader(new InputStreamReader(System.in));
        int N = Integer.parseInt(r.readLine());
        HashSet<String> set = new HashSet<>();
        String p = r.readLine();
        set.add(p);
        for(int i = 1; i < N; i++) {
            String c = r.readLine();
            if(set.contains(c) || p.charAt(p.length() - 1) != c.charAt(0)) {
                System.out.println("Player " + (i % 2 + 1) + " lost");
                return;
            }
            set.add(c);
            p = c;
        }
        System.out.println("Fair Game");
    }
}

File: wk4/cpp/shiritori.cpp
// problem link: https://open.kattis.com/problems/shiritori

#include <bits/stdc++.h>
using namespace std;

int main() {
    ios::sync_with_stdio(false); cin.tie(0); cout.tie(0);
    
	int N;
	cin >> N;
	string p;
	getline(cin, p);
	cin >> p;
	unordered_set<string> s;
	s.insert(p);
	for(int i = 1; i < N; i++) {
	    string t;
	    cin >> t;
	    if(s.count(t) || t[0] != p[p.size() - 1]) {
	        cout << "Player " << i % 2 + 1 << " lost";
	        return 0;
	    }
	    s.insert(t);
	    p = t;
	}
	cout << "Fair Game";
	    
	return 0;
}